src/api.py

```diff
@@ -40,37 +49,52 @@ def signal_handler(sig, frame):
     tornado.ioloop.IOLoop.instance().add_callback_from_signal(shutdown)
 
 
-def handle_lost_worker_node(worker_node, db_resolver):
-    # type: (dict, DatabaseResolver) -> None
-    logging.critical("Grader {} executing {} went offline unexpectedly".format(str(worker_node.get(db_key.ID)),
-                                                                               worker_node.get(db_key.RUNNING_JOBS)))
-    for job_id in worker_node.get(db_key.RUNNING_JOBS):
-        job = db_resolver.get_grading_job_collection().find_one({db_key.ID: ObjectId(job_id)})
-        db_resolver.get_grading_run_collection().update_one({db_key.ID: ObjectId(job.get(db_key.GRADING_RUN))},
-                                                            {"$inc": {db_key.STUDENT_JOBS_LEFT: -1}})
-        # TODO what if this was the last student job? Post processing might need to be scheduled
+def handle_lost_worker_node(worker_node):
+    running_job_id = worker_node.get(db_key.RUNNING_JOB)
+    worker_id = str(worker_node.get(db_key.ID))
+    logger.critical("Worker with hostname {} and id {} executing {} went offline unexpectedly".format(
+        worker_node.get(db_key.WORKER_HOSTNAME), worker_id, running_job_id))
+
+    if running_job_id is None:
+        return
+
+    # Make a fake update request on behalf of the dead worker so that the job is marked as failed and the rest of the
+    # jobs for the grading run can be handled as expected and scheduled in the right order
+    http_client = httpclient.HTTPClient()
+    res = {api_key.JOB_ID: running_job_id, api_key.SUCCESS: False,
+           api_key.RESULTS: [{"result": "Worker died while executing this job"}],
+           api_key.LOGS: {"logs": "No logs available for this job since the worker died while executing this job"}}
+    update_request = httpclient.HTTPRequest(
+        "http://localhost:{}{}/{}".format(PORT, GRADING_JOB_ENDPOINT, worker_id),
+        headers={api_key.AUTH: cluster_token},
+        method="POST", body=json.dumps(res))
```

Comment:
Instead of doing this, could we spawn an async tornado task to handle the specific function? Seems a bit cleaner to keep internal API requests internal. We could change the internal call to include failure information.

Reply:
There is no failure information available in this situation. This would happen when a worker node dies while executing a job or the grader crashes without sending an update on the job status or what happened. Were you referring to such failure message?

An async tornado task to send the POST request to itself (like it is doing now)?
Or the async task should make an internal call to the respective handler for this request?

Reply:
Async task to make an internal call to the respective handler. I would understand if this was a unit test testing the highest level of functionality but since it is the normal use case, it feels a bit awkward to post a HTTPRequest to internally for something that we expect to happen.

Reply:
Yeah, I spend a ridiculous amount of time trying to do that. But you can only make a call to a handler internally from another handler. The reason being that the handler requires a tornado request object to even create the handler object. We will need to mock the connection of the request object etc.

Reply:
Alright. I am going to give it one more try at actually refactoring that section of code. It is hard because the functionality is very tightly coupled with APIHandler class.

Reply:
Fixed in PR #32